accepted/future-releases/extension-types/feature-specification.md

```diff
@@ -1179,6 +1179,119 @@ all members of `V1, .. Vk` that are not redeclared by a declaration in _DV_
+introduced which makes `E<S1 .. Sk>` a subtype of `E<T1 .. Tk>` if
+`Sj` is a subtype of `Tj` for each `j` in 1 .. k.
+
+*For example, `E<int>` is a subtype of `E<Object>` because `int` is a
+subtype of `Object`, also in the case where `E` is an extension type.*
+
+The type function `futureValueType` is updated as follows:
+
+- `futureValueType(S?) = futureValueType(S)`, for any `S`.
+- `futureValueType(Future<S>) = S`, for any `S`.
+- `futureValueType(FutureOr<S>) = S`, for any `S`.
+- `futureValueType(void) = void`.
+- `futureValueType(dynamic) = dynamic`.
+- `futureValueType(E) = S` if `E` is an extension type
+  that implements `Future<S>`.
+- Otherwise, `futureValueType(S) = Object?`, for any `S`.
+
+The Dart 1 algorithm which is used to compute the standard upper bound of
+two distinct interface types will work in the same way as it does today,
+albeit on a different superinterface graph:
+
+We introduce an interface type `Any` which declares `operator ==`,
+`hashCode`, `runtimeType`, `noSuchMethod`, `toString` with the same
+signatures as `Object` does today, and `Object` as well as `Null` have
+`Any` as a direct superinterface.
+
+`Any` is sealed. *That is, we cannot create additional direct subtypes of
+`Any` This is required for the soundness of the rule below about `Object?`
+being a supertype of `Any`.*
+
+*This basically means that we eliminate the anomaly which is created by
+saying that `Object` declares those members, but `Null` "has them too,
+by magic". It also implies that all paths from the top type via `Object`
+are now one step longer, which is significant during the computation of the
+standard upper bound.*
+
+For subtyping, we specify that every type is a subtype of `void`, of
+`dynamic`, and of `Any`.
+
+Moreover, we specify that `Any` is a subtype of `Object?`. *This implies
+that we collapse `Any` and `Object?`, which is basically the same thing as
+promising that `Any` will never have other subtypes than `Null` and
+`Object`. Other types (type variables and extension types) will always be
+populated by objects which are of type `Object` or `Null`, which means that
+it is sound to claim that `Object?` and `Any` are "the same type" (mutual
+subtypes).*
+
+We specify that `NonNull(Any)` is `Object`. *This is sound, and it is
+needed in order to allow `Any` to be used as a more concise way to write
+`Object?`, which is presumably quite valuable.*
+
+
+### Changes to functions
+
+Let `f` be a function with declared return type `R` whose body is `async`.
+
+An existing rule states that it is a compile-time error if `R` is not a
+supertype of `Future<T>` for any `T`.
+
+This is changed as follows: If `R` is an extension type that implements
+`Future<R1>` for some `R1` which is also the representation type of `R`
+then let `S` be `Future<R1>`. Otherwise let `S` be `R`. Next, it is a
+compile-time error if `S` is not a supertype of `Future<T>` for any `T`.
```

Comment:
This should match the rule change above.

Here, we can go with:
> A return type `T` of an `async` function is a valid async function return type iff:
> * `T` is an extension type with representation type `R` and `R` is valid async function return type, or
> * `Future<Never>` \<: `T`
> 
> It's a compile-time error if the return type is not a valid async function return type.


We should perhaps just define `futureValueType` to be able to have no result, and consider that a compile time error. Rather than defining this rule which effectively amounts to "futureValueType must be well-defined for the type" and then define `futureValueType` separately.

Using the partial future value type would be something like:
> The function *futureValueType*: *type* &rarr; *type*<sub>&bot;</sub>
> is a partial function that attempts to return a type `T` for a type `S` such that `Future<T> is S` must be true,
> or which returns &bot; when such a type cannot be derived.
>
> The function is defined by the following cases, in priority order:
> * *futureValueType*(`FutureOr<T>`) = `T`
> * *futureValueType*(`Future<T>`) = `T`
> * *futureValueType*(`T?`) = *futureValueType*(`T`)
> * *futureValueType*(`E`) = *futureValueType*(`R`) if `E` is an extension type with representation type `R`
> * *futureValueType*(`dynamic`) = `dynamic`
> * *futureValueType*(`void`) = `void`
> * *futureValueType*(`T`) = `Object?` if `Future<Never>` \<: `T` (aka if **TOP**(`T`) or **OBJECT**(`T`)).
> * *futureValueType*(`T`) = &bot;

Then we can say:
> Let *V* be *futureValueType*(`R`) where `R` is the declared function type of the async function.
> It's a compile-time error if *V* is &bot;.
> Otherwise ... analyze the body using *V* as target type for returns.

Reply:
Closing, this text has been deleted.